src/pcp.rs

```diff
@@ -135,65 +199,269 @@ pub trait ValueParam: Clone + Debug {
     fn query_rand_len(&self) -> usize;
-}
 
-/// A value. Implementations of this trait specify an arithmetic circuit that determines whether an
-/// input of this type is valid.
-pub trait Value: Sized + Eq + Clone + Debug {
-    /// The finite field used for this type.
-    type Field: FieldElement;
+    /// Generate a proof of an input's validity.
+    ///
+    /// # Parameters
+    ///
+    /// * `input` is the input.
+    /// * `prove_rand` is the prover' randomness.
+    /// * `joint_rand` is the randomness shared by the prover and verifier.
+    #[allow(clippy::needless_range_loop)]
+    fn prove(
+        &self,
+        input: &[Self::Field],
+        prove_rand: &[Self::Field],
+        joint_rand: &[Self::Field],
+    ) -> Result<Vec<Self::Field>, PcpError> {
+        let mut prove_rand_len = 0;
+        let mut shim = self
+            .gadget()
+            .into_iter()
+            .map(|inner| {
+                let inner_arity = inner.arity();
+                if prove_rand_len + inner_arity > prove_rand.len() {
+                    return Err(PcpError::Prove(format!(
+                        "short prove randomness: got {}; want {}",
+                        prove_rand.len(),
+                        self.prove_rand_len()
+                    )));
+                }
 
-    /// Parameters used to construct a value of this type from a vector of field elements.
-    type Param: ValueParam;
+                let gadget = Box::new(ProveShimGadget::new(
+                    inner,
+                    &prove_rand[prove_rand_len..prove_rand_len + inner_arity],
+                )?) as Box<dyn Gadget<Self::Field>>;
+                prove_rand_len += inner_arity;
 
-    /// Constructs an instance of a value from an additive share of the input (`data`) and type
-    /// parameter (`param`). For some types, the validity circuit may depend on the total number of
-    /// shares (`num_shares`).
-    fn new_share(
-        data: Vec<Self::Field>,
-        param: &Self::Param,
-        num_shares: usize,
-    ) -> Result<Self, PcpError>;
+                Ok(gadget)
+            })
+            .collect::<Result<Vec<_>, PcpError>>()?;
+
+        // Create a buffer for storing the proof. The buffer is longer than the proof itself; the extra
+        // length is to accommodate the computation of each gadget polynomial.
+        let data_len = (0..shim.len())
+            .map(|idx| {
+                shim[idx].arity() + shim[idx].degree() * (1 + shim[idx].calls()).next_power_of_two()
+            })
+            .sum();
+        let mut proof = vec![Self::Field::zero(); data_len];
+
+        // Run the validity circuit with a sequence of "shim" gadgets that record the value of each
+        // input wire of each gadget evaluation. These values are used to construct the wire
+        // polynomials for each gadget in the next step.
+        let _ = self.valid(&mut shim, input, joint_rand, 1);
```

Comment:
You're right. I think it might have if you didn't have `let _ =` (which I like, because it explicitly tells the reader "this returns a value and we're intentionally ignoring it"). Might be worth filing an issue on clippy.